### ISSUE TITLE
Possible memory leak when running hashagg

### ISSUE THREAD

**Issue by ppggff** [greenplum-db/gpdb#7787]
There are 3 possible problems:

#### 2. After using serial/deserial/combine function for multi-phase aggregation, then there are some changes:
 * instead of using type byte(oid=17), new hashagg uses internal type(oid=2281, INTERNALOID) for passing value between process
 * combine function such as `numeric_avg_combine` will work on a internal type transValue, maybe create a new one if last one is NULL

Then, with these changes, following step may cause memory leak:
 1. run combine functions multiple time, create a lot of internal type values
 2. no room for new one, then `spill_hash_table` write these internal type values to disk
 3. goto step 1 for rest tuples

These steps will cause memory leak because the internal type is special: it stores data in memory (`aggcontext`), but set `typeByVal` to `true` in `pg_type`. It looks like a "ByValue" type as long as the related memory context exist. Then, the "hybrid hash" table treats it as "ByValue" type, `spill_hash_table` writes it's value (actually the pointer for "real" data) to file, then, combine function create a new one because last value is cleared to NULL after spilling. All "real" data is always in `aggcontext`.

Possible fix: use serial/deserial function in "hybrid hash" table for spilling, write the "real" data, free it's memory.

**Comment by gaos1** [comment]
### Greenplum version or build
6X_STABLE
### OS version and uname -a
centos 7.5
### Expected behavior
TPCDS 100GB dataset query 67a can pass
### Actual behavior
the query failed with out of memory error
### Step to reproduce the behavior
1 1 master and 8 segments, without standby and mirror
2 each segments has 2 core and 8GB memory
3 use ORCA 
4 run TPCDS 67a query. might need to tune some parameters so that the memory usage of AggContext can increase, and HashAgg needs to spill data to disk.

**Comment by adam8157** [comment]
```
$ git tag --contains bd1693e89e7e6c458d0563f6cc67a7c99a45251a
REL9_6_0
REL9_6_1
```

**Comment by adam8157** [comment]
About issue 2, Greenplum might do write actual data to file based on the binding (for sure the memory is not freed), if so the fix would be easy, will take a try. Cc @gaos1

**Comment by adam8157** [comment]
@ppggff About issue 2, check this commit https://github.com/postgres/postgres/commit/f8ace5477ef9731ef605f58d313c4cd1548f12d2

My understanding, INTERNALOID is a dummy type, it doesn't matter as long as the serialize/deserialize functions take it as the right type, those combine or "hybrid hash" functions should not see any INTERNALOID actually.

Have you tried to reproduce with the PG's commit bd1693e merged?

**Comment by ppggff** [comment]
@adam8157 Yes, there still is memory leak (because of issue 2) after merging PG's commit bd1693e.

For issue 2, the INTERNALOID type is OK for PG because PG doesn't spill hash table to file.
In GPDB, spilling hash table to file means to free memory used by hash table. But the nature of INTERNALOID type make this operation "failed", because spilling INTERNALOID type only write memory address to file, not the real data.

**Comment by adam8157** [comment]
I got the idea, two issues:
1, AggState is palloc()'ed at `aggcontext`, not the mpool `group_buf`, but we check the mpool to decide if to spill, it might OOM because not spilling.
2, Even it spills, the transition datatype for all these aggregates is declared as INTERNAL, which is a pointer, writing the pointers to file and freeing the `group_buf` doesn't relax the memory usage, might OOM.

**Comment by adam8157** [comment]
Current status:

Problem 1: merged by #7967 
Problem 3: we do reset the aggcontext in some places, waive it for now.
Problem 2: I opened a new branch working on it, https://github.com/adam8157/gpdb/tree/serialize_aggstates

**Comment by d** [comment]
> Problem 2: I opened a new branch working on it, https://github.com/adam8157/gpdb/tree/serialize_aggstates

TL;DR have we gotten a minimal repro yet? This appears to be an executor problem, we shouldn't need the entire DS suite. In fact, we shouldn't even need the entire query 67 to exercise the hash agg in question. Let's not dive into code before we know how to quickly get the executor to exhibit the memory overuse. (I'm saying this because I took a quick glance at the WIP branch, and it looks like it won't fix the problem)

When working on these "TPC-DS query X" issues, we really should get to a minimal repro, so that we can get TPC-DS completely out of our way. Keep in mind those are likely executor issues, which very likely can be exercised by much less sophisticated query plans. Simpler plans also lead to more observable execution so that we have more tools under our belt (dtrace, Xcode memory instruments, various Linux tracing tools).

Without a proper minimal repro, the only way you define success is to run a wild-ass huge suite, wait for an hour to see if there's an OOM ...

@adam8157 and other great people working on the OOM issues

Also, @ppggff do you have a minimal repro handy? Thank you for reporting this potential issue to us.

Thanks,
Jesse

**Comment by adam8157** [comment]
@ppggff if no minimal reproducer, could you please run another round with the patch from my branch applied? Thanks.

I updated the WIP branch, serialized the aggstates into the file, which would help but I'm not sure if it's the root cause.

**Comment by ppggff** [comment]
@adam8157 Thanks a lot, we will try it later this week.

**Comment by adam8157** [comment]
If possible, try https://github.com/adam8157/gpdb/tree/tpcds_oom, which is also "trying" to fix #7984

**Comment by ppggff** [comment]
@adam8157 We tried https://github.com/adam8157/gpdb/tree/serialize_aggstates with tpcds query 67a, it emitted error with following stack:

```
#0  enlargeStringInfo (needed=-4, str=0x7ffe74230cf0) at stringinfo.c:270
#1  appendBinaryStringInfo (str=str@entry=0x7ffe74230870, data=data@entry=0x7f0289ed5cec, datalen=-4) at stringinfo.c:229
#2  0x0000000000a0eae9 in numeric_avg_deserialize (fcinfo=0x7ffe74230cf0) at numeric.c:3319
#3  0x00000000007657cb in agg_hash_reload (aggstate=aggstate@entry=0x29d1de0) at execHHashagg.c:1982
#4  0x0000000000766fdc in agg_hash_next_pass (aggstate=aggstate@entry=0x29d1de0) at execHHashagg.c:2258
#5  0x0000000000737888 in ExecAgg (node=node@entry=0x29d1de0) at nodeAgg.c:1184
#6  0x000000000071df08 in ExecProcNode (node=0x29d1de0) at execProcnode.c:1085
#7  0x00000000007431e8 in ExecMaterial (node=node@entry=0x29d1868) at nodeMaterial.c:138
#8  0x000000000071df28 in ExecProcNode (node=node@entry=0x29d1868) at execProcnode.c:1077
#9  0x000000000075e2b9 in init_tuplestore_state (node=node@entry=0x2956620) at nodeShareInputScan.c:76
#10 0x000000000075ea89 in ExecShareInputScan (node=node@entry=0x2956620) at nodeShareInputScan.c:172
#11 0x000000000071de88 in ExecProcNode (node=node@entry=0x2956620) at execProcnode.c:1117
#12 0x0000000000760c58 in completeSubplan (subplan=0x2956620) at nodeSequence.c:85
#13 ExecSequence (node=node@entry=0x29561a0) at nodeSequence.c:101
#14 0x000000000071e068 in ExecProcNode (node=node@entry=0x29561a0) at execProcnode.c:982
#15 0x000000000074cd48 in ExecSort (node=node@entry=0x2955998) at nodeSort.c:185
#16 0x000000000071df18 in ExecProcNode (node=node@entry=0x2955998) at execProcnode.c:1081
#17 0x000000000075ca3c in execMotionSender (node=0x2955278) at nodeMotion.c:295
#18 ExecMotion (node=node@entry=0x2955278) at nodeMotion.c:252
#19 0x000000000071de98 in ExecProcNode (node=node@entry=0x2955278) at execProcnode.c:1113
#20 0x00000000007157f5 in ExecutePlan (estate=estate@entry=0x2954e48, planstate=0x2955278, operation=operation@entry=CMD_SELECT, sendTuples=sendTuples@entry=0 '\000', numberTuples=numberTuples@entry=0,
    direction=direction@entry=ForwardScanDirection, dest=dest@entry=0x7f028a465768) at execMain.c:2940
#21 0x000000000071643b in standard_ExecutorRun (queryDesc=0x2983f48, direction=ForwardScanDirection, count=0) at execMain.c:962
#22 0x0000000000978608 in PortalRunSelect (portal=0x2960bd8, forward=<optimized out>, count=0, dest=<optimized out>) at pquery.c:1149
#23 0x000000000097a21f in PortalRun (portal=portal@entry=0x2960bd8, count=count@entry=9223372036854775807, isTopLevel=isTopLevel@entry=1 '\001', dest=dest@entry=0x7f028a465768,
    altdest=altdest@entry=0x7f028a465768, completionTag=completionTag@entry=0x7ffe74231a90 "") at pquery.c:990
#24 0x0000000000972bbd in exec_mpp_query (
    query_string=query_string@entry=0x295cbda "with results as\n(     select i_category ,i_class ,i_brand ,i_product_name ,d_year ,d_qoy ,d_moy ,s_store_id\n", ' ' <repeats 18 times>, ",sum(coalesce(ss_sales_price*ss_quantity,0)) sumsales\n", ' ' <repeats 12 times>, "from sto"..., serializedQuerytree=serializedQuerytree@entry=0x0, serializedQuerytreelen=serializedQuerytreelen@entry=0,
    serializedPlantree=serializedPlantree@entry=0x295cfda "(\265/\375`\351", <incomplete sequence \367\225\267>, serializedPlantreelen=serializedPlantreelen@entry=5884,
    serializedParams=serializedParams@entry=0x0, serializedParamslen=serializedParamslen@entry=0, serializedQueryDispatchDesc=serializedQueryDispatchDesc@entry=0x295e6d6 "(\265/\375`\253\031\225)",
    serializedQueryDispatchDesclen=serializedQueryDispatchDesclen@entry=1340) at postgres.c:1376
#25 0x0000000000977b4f in PostgresMain (argc=<optimized out>, argv=argv@entry=0x27f8038, dbname=0x27bdde8 "tpcds_column_3000g_part2_gpadmin", username=<optimized out>) at postgres.c:5443
#26 0x000000000053a101 in BackendRun (port=0x2818b30) at postmaster.c:5037
#27 BackendStartup (port=0x2818b30) at postmaster.c:4697
#28 ServerLoop () at postmaster.c:2027
#29 0x0000000000871aa8 in PostmasterMain (argc=argc@entry=5, argv=argv@entry=0x27b7350) at postmaster.c:1557
#30 0x000000000053c34b in main (argc=5, argv=0x27b7350) at main.c:245
```

Do you need more info?

**Comment by jianlirong** [comment]
Any update for this issue?

I tried to find a minimal reproducer these days, and made some progress as follows.

I still used the TPC-DS dataset, 100GB scale, 8 segments, each is launched inside a VM with 4C (cpu) 8G (mem).

Following is the reproducer:

```sql
CREATE TABLE tmp_cr AS (
SELECT d_year AS cs_sold_year, cs_item_sk,
cs_bill_customer_sk cs_customer_sk,
cs_quantity,
cs_wholesale_cost,
cs_sales_price
FROM catalog_sales                  
   left join catalog_returns on cr_order_number=cs_order_number and cs_item_sk=cr_item_sk
   join date_dim on cs_sold_date_sk = d_date_sk
   where cr_order_number is null)
DISTRIBUTED RANDOMLY; 

ANALYZE tmp_rc;

SELECT cs_sold_year, cs_item_sk, cs_customer_sk, SUM(cs_quantity), SUM(cs_wholesale_cost), SUM(cs_sales_price) FROM tmp_cr GROUP BY cs_sold_year, cs_item_sk, cs_customer_sk limit 10;
```
The last **SELECT** statement would cause :

```sql
ERROR:  Canceling query because of high VMEM usage. Used: 7375MB, available 817MB, red zone: 7372MB (runaway_cleaner.c:189)  (seg4 slice2 192.168.108.24:40000 pid=5739) (runaway_cleaner.c:189)
```
I ran another test:

```
gpadmin=# SELECT COUNT(*) FROM (SELECT 1 FROM tmp_cr GROUP BY cs_sold_year, cs_item_sk, cs_customer_sk) tmp_tbl;
   count   
-----------
 128861352
(1 row)
```

Base on the above test results, I suspected that, as long as the number of the HashAgg group (distinct combination of cs_sold_year, cs_item_sk, cs_customer_sk) is large enough, this issue could be triggered.

**Comment by adam8157** [comment]
Current status:
1, https://github.com/adam8157/gpdb/tree/tpcds_oom is updated. The serialization and deserialization work now.
2, TPC-DS 67 works locally, but I don't have that big dataset loaded.
3, we are about to try the WIP branch on perf CI, on which we reproduced OOM before.
